src/Microsoft.Diagnostics.Monitoring.StartupHook/ProfilerResolver.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+
+using System.Runtime.InteropServices;
+using System.IO;
+using System.Reflection;
+using System;
+using Microsoft.Diagnostics.Tools.Monitor.Profiler;
+using System.Collections.Generic;
+
+namespace Microsoft.Diagnostics.Monitoring.StartupHook
+{
+    internal static class ProfilerResolver
+    {
+        private static readonly Lazy<string> s_profilerModulePath = new Lazy<string>(() => Environment.GetEnvironmentVariable(ProfilerIdentifiers.EnvironmentVariables.ModulePath) ?? string.Empty);
+        private static readonly Lazy<bool> s_profilerModuleExists = new Lazy<bool>(() => File.Exists(s_profilerModulePath.Value));
+
+        private static readonly HashSet<Assembly> s_registeredAssemblies = new HashSet<Assembly>();
+        private static readonly object s_registeredAssembliesLocker = new();
+
+        public static void InitializeResolver(Type type)
+        {
+            if (!s_profilerModuleExists.Value)
+            {
+                throw new FileNotFoundException(s_profilerModulePath.Value);
+            }
+
+            Assembly assembly = type.Assembly;
+            lock (s_registeredAssembliesLocker)
+            {
+                if (s_registeredAssemblies.Contains(assembly))
+                {
+                    return;
+                }
+
+                s_registeredAssemblies.Add(assembly);
```

Comment:
nit: you can combine your Add & Contains call into one. Add returns true if it's the first add, false otherwise.